Add drag support for dockwidgets sharing same position

goanpeca opened the pull request:
## Description

This related to dockwidgets that are tabified. Dockwidgets, when tabified, are put inside a QTabBar, but there is no access method to this bar, so to get it and process the event an eventFilter plus some search is needed. I added all the logic inside the event filter class.. but not sure if it is the best approach.

I attach a [video](http://youtu.be/wfvJupX4T5U) so you get the idea.

Nodd commented:
Nice ! Is it you playing the guitar on the video ? :smile: 
What would you put in the contextual menu ? Entries for closing/setfloating/maximizing ? I think it would be helpful only it the titlebars are hidden as per https://github.com/spyder-ide/spyder/pull/2345#issuecomment-96438578

goanpeca commented:
Hehe, no, the software I use puts the audio in, but i do play the guitar actually :-)

Now ... well I am considering two things.. if ALL widgets have a corner menu (most do on the upper right corner...) then ybe the options should go there... or also in the tab if right clicking...

I guess is natural to try to right click on a tab...

Nodd commented:
It's natural to right click but I'd expect to see options about tab management, not about the content of the page. I don't know, maybe people would use it anyway...

goanpeca commented:
Yes that is what I want, to put tab related stuff... move to right, move to right end..., move to left, move to left end..., close tab, float panel... etc

Nodd commented:
Honnestly I'm not sure the menu is worth it. In FIrefox I regularely have more than 50 tabs opened and it's a mess to manage, but in spyder I have at most 5 dockpanels at the same place, so actions like move to left or move to end are somewhat useless.

goanpeca commented:
I see what you mean, then this is almost ready :p

ccordoba12 commented:
First let me say that this is terrific addition!! Right now the way to reorganize dock widgets is to undock and dock them again, so this improves things by an order of magnitude!!

I also agree with @Nodd, I think adding a contextual menu is too much. Let's leave things simple for now :-)

goanpeca commented:
Ok, I added some docstrings and is ready for a thorough revision ;-) @ccordoba12

goanpeca commented:
@ccordoba12 this one is good to go, and easier to check (than the editor ine... that is still WIP)

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -124,6 +126,14 @@ def __init__(self, *args, **kwargs):
 
+        # Needed for the installation of the event filter
+        self.title = args[0]
+        self.main = args[1]
```
Where is `args` coming from?

goanpeca replied:
when create_dockwidget is called the args passed are the title and the parent (the main window in this case)

I agree is not so clear, but I am not so sure how to make it better...

ccordoba12 replied:
But we have `get_plugin_title` and `self.main` already. I don't get it ;-)

goanpeca replied:
But those things belong to the plugin, not to the dockwidget,

ccordoba12 replied:
I see. Then you need to redefine the __init__ method of `SpyderDockWidget` to accept the right args, i.e. something like

```python
def __init__(self, title, main, **kwargs):
    super(SpyderDockWidget, self).__init__(**kwargs)
    self.title = title
    self.main = main
    ...
```

goanpeca replied:
Fixed

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -124,6 +126,14 @@ def __init__(self, *args, **kwargs):
+        # Needed for the installation of the event filter
+        self.title = args[0]
+        self.main = args[1]
+        self.bar = None
```
wouldn't `dock_bar` be better, instead of `bar` here?

goanpeca replied:
bar here refers to tabbar... so maybe I should change it

ccordoba12 replied:
But it's the `tabbar` that holds dockwidgets, right?

So another possibility would be `dock_tabbar` :-)

goanpeca replied:
ok sure :)

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -131,6 +141,118 @@ def closeEvent(self, event):
+        self.end_tab = None
+
+    # Helper methods
+    def _get_widget(self, index):
```
Why not call this `_get_plugin` instead?

ccordoba12 replied:
I say it because I assume this is returning a `SpyderPlugin` instance, right?

goanpeca replied:
Well.. we should change the variable in spyder.py that is called widgetlist (other PR) ;-) ... and yes I agree is better with plugins...

ccordoba12 replied:
Ok

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -131,6 +141,118 @@ def closeEvent(self, event):
+    def tab_pressed(self, event):
+        """Method called when a tab from a QTabBar has been pressed."""
+        if event.button() == Qt.LeftButton:
+            self.start_tab = self.tabbar.tabAt(event.pos())
```
Do you want to change the mouse icon here too?

ccordoba12 replied:
Ups, sorry, it's below :-)

goanpeca replied:
Yep. only when it starts to move ;-)

goanpeca commented:
@ccordoba12 added corrections.

With regard to the question of the dockwidget... `SpyderPlugin` has a method create_dockwidget that uses `SpyderDockwidget` but the methods are from the plugin (SpyderPlugin) not from the widget, that is why I need to store those values when an instance if `SpyderDockwidget` is created...

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
+    def tab_moved(self, event):
+        """Method called when a tab from a QTabBar has been moved."""
+        # If the left button isn't pressed anymore then return
+        if not event.buttons() & Qt.LeftButton:
```
Why not a regular `and`, instead of `&` here?

goanpeca replied:
hmm True, proabably a leftover form the C++ code I took as example

goanpeca replied:
No I checked... buttons is an Or combination so I need the Binary And there...

http://doc.qt.io/qt-4.8/qmouseevent.html#buttons

ccordoba12 replied:
Ok, thanks.

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
+        self.moving = False
+
+    def move_tab(self, start_tab, end_tab):
+        """Move a tab from a start to and end position ."""
```
There's an unnecessary blank after `end position`

goanpeca replied:
Got it

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -124,6 +217,14 @@ def __init__(self, *args, **kwargs):
+        self.dock_tabbar = None
+
+        # Update the QTabBar everytime dockwidget visibility changes
+        self.visibilityChanged.connect(self.install_tab_event_filter)
```
Why is this necessary?

goanpeca replied:
When you move a dockwidget to a new position (location that was not existing before) this event gets triggered so I can add the filter to the QTabBar that gets created.

Was the only way I could think of so far...

goanpeca replied:
Also it even though this operation is triggered often... the filter is only installed once per QTabBar, so it should not impact usability. I hope there was a more intuitive and direct way offered by Qt, but I guess dockwidgets are meant to be used as extra panels in an application with a main  widget.. ;-)

goanpeca replied:
Oops I should have used `dockLocationChanged`

Will adapt the coed

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
+        self.moving = False
+
+    def move_tab(self, start_tab, end_tab):
+        """Move a tab from a start to and end position."""
```
`to an end`, not `to and end` :-)

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,97 @@ def get_icon(self):
+        end_plugin = self._get_plugin(end_tab)
+
+        start = plugins.index(start_plugin)
+        end = plugins.index(end_plugin)
```
What about `start_idx` and `end_idx` for these variable names? I think they should be clearer ;-)

goanpeca replied:
Sure I can change...

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -119,11 +212,21 @@ class SpyderDockWidget(QDockWidget):
-    def __init__(self, *args, **kwargs):
-        super(SpyderDockWidget, self).__init__(*args, **kwargs)
+    def __init__(self, title, parent):
+        super(SpyderDockWidget, self).__init__(title, parent)
```
Have you tried this in PyQt5? Is it working correctly there?

goanpeca replied:
Not really, but I do not see why it would not work with PyQt5

goanpeca replied:
I will check later

ccordoba12 replied:
Sorry, problems with PyQt5 only happen on mixins. So, there shouldn't be any problem with this one :-)

goanpeca commented:
I renamed variables to be more consistent with the other PR (editor draggable tabs) and added support to call context menus if needed. For the moment the methods are empty, but I added them in case they are useful in the future.

I also included the cursor fix so that it compensates when the moved tabs have different size.

I made suggested corrections and everything now is working as expected :-), so if @ccordoba12 can check and approve, this one is ready.

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
+
+    def _get_plugins(self):
+        """
+        Get a list of all the plugins references in the QTabBar to which this
```
Please change this to `Get a list of all plugin references in the QTabbar ...`

goanpeca replied:
Sure :+1:

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
+
+    def show_nontab_menu(self):
+        """Show the context menu assigned to nontabs section."""
+
```
Aren't these two last methods missing a `pass`

goanpeca replied:
I could add them, but the docstring replaces the need for a `pass`... I can add... or leave it like it is

ccordoba12 replied:
Please add `pass` to them. It feels weird without it :-)

goanpeca replied:
Done

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
+        event_type = event.type()
+        if event_type == QEvent.MouseButtonPress:
+            self.tab_pressed(event)
+            return False
```
Why these return values are needed? Could you add a comment about their need and why sometimes are True and others False?

goanpeca replied:
An event filter intercepts the fired events, and it has to return True, if it propagates the events... or False if it captures the event.... 

It is the normal behavior ....

http://doc.qt.io/qt-4.8/eventsandfilters.html#event-filters

goanpeca replied:
I will add some info to the docsting though

ccordoba12 replied:
Ok, thanks!

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,136 @@ def get_icon(self):
+
+        from_index, to_index = self.from_index, self.to_index
+        if from_index != to_index and from_index != -1 and to_index != -1:
+            self.sig_tab_moved.emit(from_index, to_index)
```
Why is this signal needed if nothing connects to it?

goanpeca replied:
At this moment it does nothing, but I guess it could be useful if some behavior wants to be executed whenever a the tabs are reorganized. It is for completeness...

ccordoba12 replied:
Then please remove it :-)

When (if) it's needed, it can easily be added. Right now it's just noise and adds confusion to the code.

goanpeca replied:
Done

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -119,11 +251,21 @@ class SpyderDockWidget(QDockWidget):
+        # To track dockwidget changes the filter is installed when dockwidget
+        # visibility changes. This installs the filter on startup and also
+        # on dockwidgets that are undocked and then docked to a new location.
+        self.visibilityChanged.connect(self.install_tab_event_filter)
```
Hadn't you mentioned that you were going to use a different signal here?

goanpeca replied:
I did, but then I remembered that it did not work properly, and I revert to this one. The problem of using the other signal is that it does not work when a dockwidget is moved to a new position.

ccordoba12 replied:
Ok, no problem.

ccordoba12 commented:
Nothing more from me to say :-) Please make my suggested changes so I can start testing.

goanpeca commented:
@ccordoba12 done

ccordoba12 commented on spyderlib/plugins/__init__.py:
```diff
@@ -55,6 +57,141 @@ def get_icon(self):
+
+    def show_nontab_menu(self):
+        """Show the context menu assigned to nontabs section."""
+        pass
```
Las thing, I think I have the perfect suggestion for this entry

```python
    self.main.createPopupMenu()
```

goanpeca replied:
Ehh ? what does it do?

goanpeca replied:
http://doc.qt.io/qt-4.8/qmainwindow.html#createPopupMenu ?

ccordoba12 replied:
It shows the same context menu as when you do a right click over a toolbar.

Essentially, It shows the Help menu but with just a few entries.

goanpeca replied:
it does not work...

goanpeca replied:
Ok fixed it :), now this will only popup when clicking in an area where no tabs are present...

In this image I right clicked to the right of the `File explorer` Tab

![image](https://cloud.githubusercontent.com/assets/3627835/7445834/55482308-f1c3-11e4-8482-1ae441a62a43.png)

goanpeca commented:
@ccordoba12 you want to get the same context menu when right clicking on a tab as well??

I have two methods in this PR, one is `show_tab_menu` (right click on a tab) and `show_nontab_menu` (right click on a QTabBar empty space, generally to the right of the tabs)

ccordoba12 commented:
I'd say yes, for now. But leave the two methods, so if we come up with a better context menu for tabs on the future, we can change it easily :-)

goanpeca commented:
Ready for testing @ccordoba12

ccordoba12 commented:
Working great, even in PyQt5!

Nothing else than merging ;-)
